stdlib/event/gui.lua

```diff
@@ -2,42 +2,15 @@
 -- @module Gui
 -- @usage local Gui = require('stdlib/event/gui')
 
-require('stdlib/event/event')
+require('stdlib/utils/table')
```

Comment:
Redundant call as it is already included in core.

Reply:
I am pretty confused about how core actually works.  I don't know:

* Is there some meta-magic or convention that's supposed to guarantee core stuff is in the module's _ENV (that is, can/should certain things in core be treated as globals in module code)?
* Does the core metatable provide a certain API to consumers of stdlib modules?  If so, what?
* What is the correct/best way to build a new stdlib module (including the core stuff)?
* What is the "friend" API surface of core -- put another way, if I am coding a module, are there things in core I should /not/ be consuming as they are considered implementation details that might change?
* Is there official documentation for core?

That's a lot of questions so feel free to cherry-pick :)  Any insights would be appreciated.

Reply:
Core is just the core, All the stdlib classes have an __index into core using setmetatable

When core is first required it requires some files
table/string/Iter/Is (etc) and makes them global (in the case of table and string it mutates the existing global)
core.lua
```lua
local Core = {}
Core.func = function() return true end
return Core
```
data.lua
```lua
local Data = setmetatable({}, {__index = require("core")}) --require checks in package.loaded for Core if it exists it returns that value (the return return from core.lua is the value), otherwise require  the file and stick the return here
return Data
```
file.lua
local Data = require("data")
if Data.func() then print("good times") end
```

Reply:
Nexela said:
> data.lua:
> ```lua
> local Data = setmetatable({}, {__index = require("core")}) -- ...
> return Data
> ```

I guess what is confusing me is, what is the purpose of using __index here?

For example, doesn't this mean that after a factorio mod does  `Data = require('stdlib/data/data')`, then `data._module_name == "Core"`, `data.VALID_FILTER(blah) == blah.valid`, etc?  I guess that is a secondary side effect, that we are assuming won't be messed around with by stdlib consumers?

But, then, what are all those _x virtuals actually doing for the stdlib module that inherits them?  Or is that too confusing to be explainable in this context, i.e., I should just read more stdlib code if I want to grok this?